3Dmol/glviewer.js

```diff
@@ -3314,34 +3312,34 @@ $3Dmol.GLViewer = (function() {
 
                 var sync = !!($3Dmol.syncSurface);
                 if (sync) { // don't use worker, still break up for memory purposes
-
+                    var cnt = 0;
                     // to keep the browser from locking up, call through setTimeout
                     var callSyncHelper = function callSyncHelper(i) {
-                        if (i >= extents.length) {
-                            surfobj.done = true;
-                            if(surfacecallback && typeof(surfacecallback) == "function") {
-                                surfacecallback(surfid);
+                        return new Promise(function(resolve, reject) { 
+                            var VandF = generateMeshSyncHelper(type, extents[i].extent,
+                                    extents[i].atoms, extents[i].toshow, reducedAtoms,
+                                    totalVol);
+                            //complicated surfaces sometimes have > 2^16 vertices
+                            var VandFs = $3Dmol.splitMesh({vertexArr:VandF.vertices, faceArr:VandF.faces});
+                            for(var vi=0,vl=VandFs.length;vi<vl;vi++){
+                                var VandF={vertices:VandFs[vi].vertexArr,
+                                        faces:VandFs[vi].faceArr};                            
+                                var mesh = generateSurfaceMesh(atomlist, VandF, mat);
+                                $3Dmol.mergeGeos(surfobj.geo, mesh);
                             }
-                            return;
-                        }
-
-                        var VandF = generateMeshSyncHelper(type, extents[i].extent,
-                                extents[i].atoms, extents[i].toshow, reducedAtoms,
-                                totalVol);
-                        //complicated surfaces sometimes have > 2^16 vertices
-                        var VandFs = $3Dmol.splitMesh({vertexArr:VandF.vertices, faceArr:VandF.faces});
-                        for(var vi=0,vl=VandFs.length;vi<vl;vi++){
-                            var VandF={vertices:VandFs[vi].vertexArr,
-                                    faces:VandFs[vi].faceArr};                            
-                            var mesh = generateSurfaceMesh(atomlist, VandF, mat);
-                            $3Dmol.mergeGeos(surfobj.geo, mesh);
-                        }
-                        _viewer.render();
-
-                        setTimeout(callSyncHelper, 1, i + 1);
+                            resolve();
+                        })
+                        .then(function() {
+                            _viewer.render();
```

Comment:
I don't think all the extents will be guaranteed to be completed when the promise resolves.  Am I wrong? (Also should resolve with the surface id as below).